fms_extras/models/speculator.py

```diff
@@ -170,3 +170,94 @@ def forward(
             state = self.activation(self.ln[i](state))  # b n d
             out.append(self.head[i](state))  # b n v
         return torch.stack(out, dim=0)  # h b n v
+
+
+def select_inflate_dim(
+    inp: torch.Tensor, inds: torch.Tensor, dim: int = 0
+) -> torch.Tensor:
+    """
+    Takes an input of size ([...], n, [...]), with n in slot corresponding to value of dim,
+    and tensor of indices of size (a, ..., z). Using those indices we over/under sample the
+    input on dimension n, to create output tensor with size ([...], (a, ..., z), [...]).
+
+    i.e. if dim=0, inp has size (6,3,2), and inds has size (8,4), then:
+    1) max(inds) < 6
+    2) output has size (8,4,3,2)
+
+    Args:
+        inp: torch.Tensor
+            tensor of inputs
+        inds: torch.Tensor
+            tensor of indices
+        dim: int
+            dimension to sample on
+
+    Returns:
+        torch.Tensor
+            output tensor with new size ([...], (a, ..., z), [...])
+    """
+    inds_shape = inds.size()
+    inp_shape = inp.size()
+    out = inp.index_select(dim, inds.view(-1))
+    return out.view(*inp_shape[:dim], *inds_shape, *inp_shape[dim + 1 :])
+
+
+def flatten_batch(inp: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    """
+    Takes a speculator suffix tree: a bsize x n_candidates x candidate_len rectangular batch
+    of token indices, and flattens it while removing redundant tokens.
+
+    For example, given:
+
+    a b c
+    a b d
+    a e f
+
+    Tokens 'a b' in line 2 and token 'a' in line 3 are functionally equivalent to 'a b' in
+    line 1, so the flattened batch returns `a b c d e f`
```

Comment:
the example here doesn't make it very clear how the information is preserved, but I think what's described here is only the first of the three tensors returned? it might help to see this in a unit test so I can see how the other indices tensors map the 'flattened' tensor back to the source data.

Reply:
Yes exactly, this is just the first output. Unit tests have been added.